arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -86,11 +86,20 @@ static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
 
 static inline u8_t _get_num_regions(void)
 {
+#if defined(CONFIG_CPU_CORTEX_M0PLUS) || \
+	defined(CONFIG_CPU_CORTEX_M3) || \
+	defined(CONFIG_CPU_CORTEX_M4)
+	/* Cortex-M0+, Cortex-M3, and Cortex-M4 MCUs may
+	 * have a fixed number of 8 MPU regions.
+	 */
+	return 8;
```

Comment:
This makes it essential that arm_mpu.c is _not_ linked, if platform does not support MPU. This is because the function returns a constant; MPU_TYPE register is not queried at run-time, so it cannot return zero.

The __ASSERT that is added in _arm_mpu_init(), somewhat, takes care of it; it queries the DREGION (number of regions) and makes sure it is 8 for CM0+, CM3, CM4.